kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/analyze/TopicAnalyzeStats.java

```diff
@@ -0,0 +1,142 @@
+  final HllSketch uniqValues = new HllSketch();
+
+  final HourlyCounts hourlyCounts = new HourlyCounts();
+
+  static class SizeStats {
+    long sum = 0;
+    Long min;
+    Long max;
+    final UpdateDoublesSketch sizeSketch = DoublesSketch.builder().build();
+
+    void apply(byte[] bytes) {
+      int len = bytes.length;
+      sum += len;
+      min = minNullable(min, len);
+      max = maxNullable(max, len);
+      sizeSketch.update(len);
+    }
+
+    TopicAnalyzeSizeStatsDTO toDto() {
+      return new TopicAnalyzeSizeStatsDTO()
+          .sum(sum)
+          .min(min)
+          .max(max)
+          .avg((long) (((double) sum) / sizeSketch.getN()))
+          .prctl50((long) sizeSketch.getQuantile(0.5))
+          .prctl75((long) sizeSketch.getQuantile(0.75))
+          .prctl95((long) sizeSketch.getQuantile(0.95))
+          .prctl99((long) sizeSketch.getQuantile(0.99))
+          .prctl999((long) sizeSketch.getQuantile(0.999));
+    }
+  }
+
+  static class HourlyCounts {
+
+    // hour start ms -> count
+    private final Map<Long, Long> hourlyStats = new HashMap<>();
+    private final long minTs = Instant.now().minus(Duration.ofDays(14)).toEpochMilli();
+
+    void apply(ConsumerRecord<?, ?> rec) {
+      if (rec.timestamp() > minTs) {
+        var hourStart = rec.timestamp() - rec.timestamp() % (1_000 * 60 * 60);
+        hourlyStats.compute(hourStart, (h, cnt) -> cnt == null ? 1 : cnt + 1);
+      }
+    }
+
+    List<TopicAnalyzeStatsHourlyMsgCountsDTO> toDto() {
+      return hourlyStats.entrySet().stream()
+          .sorted(Comparator.comparingLong(Map.Entry::getKey))
+          .map(e -> new TopicAnalyzeStatsHourlyMsgCountsDTO()
+              .hourStart(e.getKey())
+              .count(e.getValue()))
+          .collect(Collectors.toList());
+    }
+  }
+
+  void apply(ConsumerRecord<Bytes, Bytes> rec) {
+    totalMsgs++;
+    minTimestamp = minNullable(minTimestamp, rec.timestamp());
+    maxTimestamp = maxNullable(maxTimestamp, rec.timestamp());
+    minOffset = minNullable(minOffset, rec.offset());
+    maxOffset = maxNullable(maxOffset, rec.offset());
+    hourlyCounts.apply(rec);
+
+    if (rec.key() != null) {
+      byte[] keyBytes = rec.key().get();
+      keysSize.apply(keyBytes);
+      uniqKeys.update(keyBytes);
+    } else {
+      nullKeys++;
+    }
+
+    if (rec.value() != null) {
+      byte[] valueBytes = rec.value().get();
+      valuesSize.apply(valueBytes);
+      uniqValues.update(valueBytes);
+    } else {
+      nullValues++;
+    }
+  }
+
+  TopicAnalyzeStatsDTO toDto(@Nullable Integer partition) {
+    return new TopicAnalyzeStatsDTO()
+        .partition(partition)
+        .totalMsgs(totalMsgs)
+        .minOffset(minOffset)
+        .maxOffset(maxOffset)
+        .minTimestamp(minTimestamp)
+        .maxTimestamp(maxTimestamp)
+        .nullKeys(nullKeys)
+        .nullValues(nullValues)
+        // because of hll error estimated size can be greater that actual msgs count
```

Comment:
what's hll?

Reply:
HyperLogLog structure. Here it is used to calculate approx number of uniq keys & values